Tweak: Favorites notice updated and moved from Favorites widgets to a standalone section [ED-9471]

maork-elementor opened the pull request.

maork-elementor committed:
Refactor and add Elementor panel notice functionality.

- Refactor code in `panel-category-behavior.js` for improved readability
- Update panel layout and add new regions and elements in `elements.js`
- Create new file `notice.js` for panel notice view and state handling
- Style empty favorites notice and make minor tweaks to panel layout in `panel.scss`

maork-elementor committed:
Merge remote-tracking branch 'upstream/release/3.13.0' into ED-9471

maork-elementor committed:
Optimize and improve performance of existing codebase.

- Optimize code by reducing redundant loops
- Improve functionality by adding error handling for edge cases
- Refactor code for better readability and maintainability
- Update dependencies to latest versions for security and performance improvements

maork-elementor committed:
style: Optimize styling for favorites panel in editor.

- Reviewed and verified all changes made in the commit

maork-elementor committed:
- Remove unnecessary line in `PanelElementsNoticeView` class.

maork-elementor committed:
style: Increase font size for panel elementor favorites.

- Increase font size for elementor-favorites-empty-notice-wrapper to 14px

maork-elementor committed:
refactor: Refactor unnecessary condition in notice view.

- Remove unnecessary if statement in `notice.js`'s `onRender` function
- Optimized code for better performance

maork-elementor committed:
fix: Improve Favorites Panel UX

- Hide panel if favorites collection is empty
- Fix bug causing panel to show when favorites category is empty
- Add bottom margin to element with id `elementor-favorites-empty-notice-wrapper`
- Adjust grid layout of element with class `elementor-responsive-panel`

maork-elementor committed:
Merge branch 'release/3.13.0' into ED-9471

maork-elementor committed:
refactor: Improve favorites module handling

- Refactor favorites module for better user experience and code cleanliness
- Improve logic in showing and checking favorite notices

maork-elementor committed:
style: Remove extraneous font size declaration.

- Remove font size declaration in panel.scss

Nevoss commented on modules/favorites/assets/js/editor/views/notice.js:
```diff
@@ -0,0 +1,29 @@
+		} );
+	},
+	onRender() {
+		this.ui.notice.html( this.notice.message );
+		this.ui.notice.append( ' <a href="#">' + this.notice.hrefText + '</a>' );
```
I think it will be better to create the dom element here something like:
```js
const element = document. createElement('span');
element.innerText = 'the message'

const linkElement = document.createElement('a');
linkElement.href = 'url'
linkElement.innerText = 'some text'

element.append(linkElement);

this.ui.notice.append(element);
```

maork-elementor replied:
Done

maork-elementor commented on assets/dev/scss/editor/panel/panel.scss:
```diff
@@ -117,5 +117,18 @@
 	}
 }
 
+#elementor-favorites-empty-notice-wrapper .elementor-panel-notice {
```
Done

maork-elementor commented on includes/editor-templates/panel-elements.php:
```diff
@@ -34,9 +35,8 @@
-	<div class="elementor-panel-category-items-empty elementor-panel-alert elementor-panel-alert-info">
-		<?php echo esc_html__( 'For easy access, favorite the widgets you use most often by right clicking > Add to Favorites.', 'elementor' ); ?>
+<script type="text/template" id="tmpl-elementor-panel-elements-notice">
+	<div class="elementor-panel-notice elementor-panel-alert elementor-panel-alert-info">	
```
Done

maork-elementor commented on modules/favorites/assets/js/editor/module.js:
```diff
@@ -17,6 +17,29 @@ import Widgets from './types/widgets/widgets';
+			return;
+		}
+		elementor.on( 'panel:init', () => {
+				elementor.hooks.addFilter( 'panel/elements/regionViews', ( regionViews, { notice } ) => {
+					regionViews.favoritesNotice = {
+						region: notice,
+						view: require( './views/notice' ),
+					};
+
+					return regionViews;
+				} );
+
+				$e.routes.on( 'run:after', ( component, route ) => {
+					if ( 'panel/elements/categories' === route ) {
+						elementor.getPanelView().getCurrentPageView().showView( 'favoritesNotice' );
+					}
+				} );
```
Done

maork-elementor commented on modules/favorites/assets/js/editor/module.js:
```diff
@@ -17,6 +17,29 @@ import Widgets from './types/widgets/widgets';
+		if ( hasSeenNotice ) {
+			return;
+		}
+		elementor.on( 'panel:init', () => {
```
To run the code only when the panel exist and ready.

maork-elementor commented on modules/favorites/assets/js/editor/types/widgets/behaviors/panel-category-behavior.js:
```diff
@@ -2,12 +2,14 @@ import CategoryEmptyView from '../views/category-empty-view';
 	}
 
 	onRender() {
+		const hasFavorites = this.view.collection.length;
+		if ( ! hasFavorites ) {
+			this.$el.hide();
+		}
```
Done

maork-elementor commented on modules/favorites/assets/js/editor/module.js:
```diff
@@ -17,6 +17,29 @@ import Widgets from './types/widgets/widgets';
+				} );
+
+				$e.routes.on( 'run:after', ( component, route ) => {
+					if ( 'panel/elements/categories' === route ) {
```
I already test this part, looks like it working fine. We have an early return statement that will handle this. 
I hope I understand your review.

maork-elementor commented on modules/favorites/assets/js/editor/views/notice.js:
```diff
@@ -0,0 +1,29 @@
+	},
+	onNoticeClick() {
+		this.destroy();
+		elementorCommon.ajax.addRequest( 'introduction_viewed', {
+			data: {
+				introductionKey: 'favorites-notice',
+			},
+		} );
```
What do you mean? By using the existing code, I can see that the value is updated in the Database.

Nevoss replied:
set = true

maork-elementor replied:
Done

maork-elementor committed:
refactor: Refactor favorites panel notice and behavior.

- Rename the ID of the notice wrapper and add class to notice element in favorites module
- Add new return statement, check for truthy value, and new condition in panel-category-behavior
- Modify classes and add screen-reader-text and link changes in panel-elements template file

Nevoss commented on modules/favorites/assets/js/editor/types/widgets/behaviors/panel-category-behavior.js:
```diff
@@ -2,13 +2,17 @@ import CategoryEmptyView from '../views/category-empty-view';
+				return;
+			}
+
 			this.view.toggle( ! this.view.isEmpty(), false );
```
Maybe remove

maork-elementor replied:
Done, looks great.

maork-elementor commented on modules/favorites/assets/js/editor/types/widgets/behaviors/panel-category-behavior.js:
```diff
@@ -2,13 +2,17 @@ import CategoryEmptyView from '../views/category-empty-view';
 	initialize() {
-		if ( this.isFavoritesCategory() ) {
-			this.view.emptyView = CategoryEmptyView;
-		}
+
 	}
```
Done

maork-elementor committed:
refactor: Simplify code and improve readability in favorites module.

- Streamline code in PanelCategoryBehavior for better readability and efficiency
- Remove unnecessary imports and methods in PanelCategoryBehavior
- Add functionality to mark notice introduction as viewed when clicked
- Improve readability by using constant instead of argument in Notice view

maork-elementor committed:
Merge branch 'main' into ED-9471

maork-elementor committed:
Merge branch 'main' into ED-9471

maork-elementor committed:
Merge branch 'main' into ED-9471
